### PR TITLE
Consolidate the last easy parser construction

### PR DESCRIPTION
**Pull request by nik9000**
Moves the last of the "easy" parser construction into
`RestRequest`, this time with a new method
`RestRequest#contentParser`. The rest of the production
code that builds `XContentParser` isn't "easy" because it is
exposed in the Transport Client API (a Builder) object.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/update/UpdateRequest.java [hunk 1]
```diff
@@ -717,52 +703,48 @@ public boolean detectNoop() {
...
-                    } else if (token.isValue()) {
-                        fields = Collections.singletonList(parser.text());
-                    }
-                    if (fields != null) {
-                        fields(fields.toArray(new String[fields.size()]));
-                    }
-                } else if ("_source".equals(currentFieldName)) {
-                    fetchSourceContext = FetchSourceContext.parse(parser);
+        XContentParser.Token token = parser.nextToken();
+        if (token == null) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/update/UpdateRequest.java]: Just indentation changed.

**Comment by nik9000** [comment]
@rjernst, this is another prerequisite to #22003. This one should be more obviously in line with the objective of #22003 than the last one.

My plan for the not "easy" `XContentParser` constructions is to do them in the API with an empty `NamedXContentRegistry` for 5.x and remove them in 6.0.

File: core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestRestoreSnapshotAction.java [hunk 2]
```diff
@@ -49,7 +50,11 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         RestoreSnapshotRequest restoreSnapshotRequest = restoreSnapshotRequest(request.param("repository"), request.param("snapshot"));
         restoreSnapshotRequest.masterNodeTimeout(request.paramAsTime("master_timeout", restoreSnapshotRequest.masterNodeTimeout()));
         restoreSnapshotRequest.waitForCompletion(request.paramAsBoolean("wait_for_completion", false));
-        restoreSnapshotRequest.source(request.content().utf8ToString());
+        if (request.hasContent()) {
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestRestoreSnapshotAction.java]: maybe for usages like this one we can add a helper method on `RestRequest`? Like this one:

```java
public final void applySource(Consumer<Map<String, Object>> target) throws IOException {
    if (hasContent()) {
        try (XContentParser parser = contentParser()) {
            target.accept(parser.mapOrdered());
        }
   }
}
```

Then this snippet can be replaced with:

```java
request.applySource(restoreSnapshotRequest::source);
```

and we could have a similar helper that passes a `XContentParser` to an action request?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestRestoreSnapshotAction.java]: Sure! I'll make that change.

**Comment by nik9000** [comment]
@martijnvg I made the change you asked for.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java [hunk 3]
```diff
@@ -198,83 +196,23 @@ public boolean verify() {
...
-                settings((Map<String, Object>) entry.getValue());
+                @SuppressWarnings("unchecked")
+                Map<String, Object> sub = (Map<String, Object>) entry.getValue();
+                settings(sub);
             }
         }
         return this;
     }
 
-    /**
```
Comment (imotov) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java]: This is user-facing interface and removing this calls might break applications that are using transport client to create repositories. So, I think we need to start with marking this as deprecated before removing it completely. Same in other requests.

File: core/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java [hunk 4]
```diff
@@ -20,10 +20,13 @@
 package org.elasticsearch.rest.action.admin.indices;
 
 import org.elasticsearch.action.admin.indices.rollover.RolloverRequest;
+import org.elasticsearch.action.admin.indices.shrink.ShrinkRequest;
```
Comment (imotov) [inline core/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java]: Nitpicking - unused imports were added to this and other IndexActions.

**Review by imotov — COMMENTED** [review]
Left a couple of comments.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java (hunk 3 above)
Comment (imotov) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java]: I am not totally sure about this one. I think it depends on user preferences at the moment. The transport client can be used with builders or without. All our examples are showing `client.prepareBlah(...)` version, but the non-builder version `client.blah(...)` is readily accessible to anyone who is using client and I don't remember us ever mentioning that it's internal and shouldn't be used in apps.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java]: they are both user facing like Igor said, but not sure we have been keeping bw comp on the java api between minor versions. I think it's ok to make these changes as long as we mark breaking-java and we document what the replacement is. We generally deprecate stuff that has to do with the REST layer. I think.

**Review by imotov — APPROVED** [review]
LGTM

**Comment by nik9000** [comment]
Thanks for all the reviews!

master: 749039ad4f4ff311dcaa17463dafc82d42174db1
5.x: backporting

**Pull request merged**